### ISSUE TITLE
Signature incompatible with mapping when key is a union of 2 types

### ISSUE THREAD

**Issue by spacether** [python/mypy#15663]
**Bug Report**

I am getting this unexpected error:
```
overload_mypy.py:15: error: Signature of "__getitem__" incompatible with supertype "dict"
overload_mypy.py:15: note:      Superclass:
overload_mypy.py:15: note:          def __getitem__(self, Union[KeyEnum, ADDITIONAL_PROPERTY]) -> Union[int, str]
overload_mypy.py:15: note:      Subclass:
overload_mypy.py:15: note:          @overload
overload_mypy.py:15: note:          def __getitem__(self, Literal[KeyEnum.PROP_A]) -> int
overload_mypy.py:15: note:          @overload
overload_mypy.py:15: note:          def __getitem__(self, ADDITIONAL_PROPERTY) -> str
overload_mypy.py:15: error: Signature of "__getitem__" incompatible with supertype "Mapping"
overload_mypy.py:15: note:          def __getitem__(self, Union[KeyEnum, ADDITIONAL_PROPERTY]) -> Union[int, str]
overload_mypy.py:15: note:          @overload
overload_mypy.py:15: note:          def __getitem__(self, Literal[KeyEnum.PROP_A]) -> int
overload_mypy.py:15: note:          @overload
overload_mypy.py:15: note:          def __getitem__(self, ADDITIONAL_PROPERTY) -> str
Found 2 errors in 1 file (checked 1 source file)
```

**To Reproduce**

```python
import enum
import typing
import typing_extensions

class ADDITIONAL_PROPERTY(str):
    pass

class KeyEnum(str, enum.Enum):
    PROP_A = "prop_a"

OUTPUT_BASE_TYPES = typing.Union[int, str]

class SomeDict(typing.Dict[typing.Union[KeyEnum, ADDITIONAL_PROPERTY], OUTPUT_BASE_TYPES]):
    
    @typing.overload
    def __getitem__(self, name: typing_extensions.Literal[KeyEnum.PROP_A]) -> int:
        ...
    
    @typing.overload
    def __getitem__(self, name: ADDITIONAL_PROPERTY) -> str:
        ...
    
    def __getitem__(
        self,
        name
    ):
        # dict_instance[name] accessor
        return super().__getitem__(name)
```

**Expected Behavior**

code should pass a mypy check because keys are constrained to strings or an enum
overloads are exhaustive

**Actual Behavior**

my reports errors

**Your Environment**

- Mypy version used: mypy 0.971 (compiled: yes)
- Mypy command-line flags: n/a
- Mypy configuration options from `mypy.ini` (and other config files): n/a
- Python version used: Python 3.9.17 (main, Jun  6 2023, 14:44:03) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin

**Comment by spacether** [comment]
One can get this working by building an implementation of typing.Mapping with manual definition fo al getitem overloads:
```
import enum
import typing
import typing_extensions

class KeyEnum(enum.Enum):
    PROP_A = "prop_a"
    PROP_B = "prop_b"

OUTPUT_BASE_TYPES = typing.Union[int, str]

_KE = typing.TypeVar('_KE')
_VT = typing.TypeVar('_VT')


class GenericDict(typing.Mapping, typing.Generic[_KE, _VT]):
    def __new__(cls, arg: typing.Mapping[typing.Union[_KE, str], _VT]):
        inst = super().__new__(cls)
        setattr(inst, '_dict', arg)
        return inst

    def __iter__(self):
        inst = getattr(self, '_dict')
        return inst.__iter__()

    def __len__(self):
        inst = getattr(self, '_dict')
        return inst.__len__()


class SomeDict(GenericDict[KeyEnum, OUTPUT_BASE_TYPES]):
    @typing.overload
    def __getitem__(self, name: typing_extensions.Literal[KeyEnum.PROP_A]) -> int:
        ...
    
    @typing.overload
    def __getitem__(self, name: str) -> str:
        ...
    
    def __getitem__(
        self,
        name
    ):
        # dict_instance[name] accessor
        inst = getattr(self, '_dict')
        res = inst[name]
        if name is KeyEnum.PROP_A:
            return typing.cast(int, res)
        return typing.cast(str, res)

inst = SomeDict({
    KeyEnum.PROP_A: 1,
    'addProp': 'hi'
})
print(inst)
val = inst[KeyEnum.PROP_A]
print(val)
addpro_val = inst['addProp']
print(addpro_val)
print(len(inst))
print([v for v in inst])
```
